blog/2024/preview-keycloak-test-framework.adoc

```diff
@@ -0,0 +1,42 @@
+
+For our purpose, we have https://github.com/keycloak/keycloak/blob/main/test-framework/core/src/main/java/org/keycloak/test/framework/KeycloakIntegrationTestExtension.java[implemented] these three callback classes:
+----
+BeforeEachCallback
+AfterEachCallback
+AfterAllCallback
+----
+
+After a successful test round, we've continued with a proof of concept extending features list to support multiple server modes, different databases and _WebDrivers_, clients and users setup, _SmallRye_ configuration support, _OAuthClient_ based on https://connect2id.com/products/nimbus-oauth-openid-connect-sdk[_Nimbus SDK_] (this feature is a preview only) etc. The full list of currently implemented features can be found here in the https://github.com/keycloak/keycloak/tree/main/test-framework/core/src/main/java/org/keycloak/test/framework/annotations[annotations package]. It is already present in the main branch and _Keycloak_ nightly builds.
+
+== Are you curious about where to start?
+We suggest reading the https://github.com/keycloak/keycloak/blob/main/test-framework/README.md[user guide], which will provide a basic overview of how the framework works and should be used. If this is not enough, you can also check https://github.com/keycloak/keycloak/tree/main/test-framework/examples/tests/src/test/java/org/keycloak/test/examples[test examples].
+
+For extension developers we recommend to look into an example on how to start _Keycloak_ with their custom provider: https://github.com/keycloak/keycloak/tree/main/test-framework/examples/providers[provider example], https://github.com/keycloak/keycloak/blob/main/test-framework/examples/tests/pom.xml#L75-L80[pom.xml test dependency] and https://github.com/keycloak/keycloak/blob/main/test-framework/examples/tests/src/test/java/org/keycloak/test/examples/MyCustomProviderTest.java[test example].
+
+If you find a bug, want to discuss something, or propose a new enhancement, please follow this GitHub feedback https://www.github.com/discussion#blabla[discussion link].
+
+== Coming Soon
+
+Main plan is to cover enough features for most functionalities currently available in the https://github.com/keycloak/keycloak/tree/main/testsuite[integration test suite]. Next future step shall be to introduce some new concepts, e.g. _@KeycloakTest_ annotation, which will add an option to test providers without the _Keycloak_ server running. It should be based on the _KeycloakSession_ class. Other interesting features should be https://junit.org/junit5/docs/current/user-guide/#junit-platform-suite-engine[_JUnit5_ test suites], parallel execution on a GitHub runner or enabling a Keycloak cluster setup.
```

Comment:
```suggestion
== Next steps

We already have enough capabilities in the new test framework to start migrating some tests from the old testsuite; and in fact already have our very first test migrated. We plan to migrate one package at a time from the old testsuite starting with `admin` tests, then moving on to `forms` and `oauth` packages. As we are doing this we will expand on the capabilities of the test framework.

Some features we know will be coming soon include:

* An easier way to deploy custom providers, not requiring a Maven build of the provider first
* Improved logging, making it easy to configure logging from tests as well as Keycloak
* Easy testing of OAuth and OpenID Connect, including a mock application
* Extension to allow running code on the tested server when it's not possible to easily test through only remote interfaces
```

We also have some more long term plans to deliver:

* Provider tests that can be used to easily test a provider by invoking the provider directly
* Parallel execution of tests, to take full advantage of multiple cores to reduce test execution time